torch/_library/fake_profile.py

```diff
@@ -0,0 +1,142 @@
+            )
+
+        if isinstance(out_profile, TensorMetadata):
+            return _generate_tensor_out(out_profile)
+        else:
+            return [_generate_tensor_out(t) for t in out_profile]
+
+    def _fake_kernel(*args, **kwargs):  # type: ignore[no-untyped-def]
+        for profile in op_profile:
+            if _match_args(profile.args_profile, (*args, *kwargs.values())):
+                return _generate_res(profile.out_profile)
+
+        raise MissingOpProfile(
+            f"No fake kernel was found for {op_name}, and although we have "
+            "previously registered some profiles to generate a fake kernel, "
+            f"no profiles match the given inputs: {args, kwargs}."
+        )
+
+    return _fake_kernel
+
+
+@contextlib.contextmanager
+def register_fake_profile(op_profiles: dict[str, set[OpProfile]]) -> Generator:
+    """
+    Registers a fake kernel based on the given operator profiles. This fake
+    kernel registration will override any existing fake kernel registrations.
+
+    The input is a dictionary mapping operator names to a set of operator
+    profiles, which we will use to generate fake kernels. The operator profiles
+    are a record of the input and output tensor metadata. Based on this
+    information we will match a given input to the recorded profile, and return
+    an output with the same metadata as in the recorded profile. If a profile
+    doesn't exist then an exception will be thrown.
+
+    Args:
+        op_profiles (dict[str, set[OpProfile]]): A dictionary mapping operator
+            name to a set of operator profiles from which we will generate fake
+            kernels.
+    """
+
+    libs: list[torch.library.Library] = []
+    # Stores old fake impls from custom ops declared through @custom_op
+    old_fake_impls: dict[str, Callable] = {}
+    for op_name, profiles in op_profiles.items():
+        log.warning(
+            "Registering fake profile for %s. This will override any existing "
+            "fake kernel registration.",
+            op_name,
+        )
+
+        op_name_split = op_name.split(".")
+        namespace, op_name_str = op_name_split[0], op_name_split[1]
+        op_str = f"{namespace}::{op_name_str}"
+
+        newlib = torch.library.Library(namespace, "FRAGMENT")  # noqa: TOR901
+        fake_kernel = _generate_fake_kernel(op_str, profiles)
+
+        # Check if there are any fake kernels already registered
+        if opdef := _maybe_get_opdef(op_str):
+            if opdef._abstract_fn is not None:
+                old_fake_impls[op_str] = opdef._abstract_fn
+
+        torch.library.register_fake(
+            op_str, fake_kernel, lib=newlib, allow_override=True
+        )
+        libs.append(newlib)
+
+    try:
+        yield libs
+    finally:
+        for lib in libs:
+            lib._destroy()
+
+        for op_str, old_fake in old_fake_impls.items():
+            torch.library.register_fake(op_str, old_fake, allow_override=True)
```

Comment:
I'm confused, this seems like in the regular torch.library.register_fake case, we would keep on growing the number of kernels.

Maybe we should handle the torch.library.register_fake case separately from the CustomOpDef.register_fake case.

Reply:
Ah sorry, this should only happen if we're in the CustomOpDef case. I refactored the code and added comments to demonstrate that better